### DISCUSSION TITLE
Restart Concurrent Subscription Worker

### DISCUSSION THREAD

**Discussion by tehkyle** [ravendb/ravendb#14699] [Q&A]
Is it possible to restart cuncurrent subscription workers? Prior to the concurrent addition I was just removing and recreating the subscription, but would prefer not to cause all other workers to restart in the concurrent setup.

**Comment by ayende** [comment]
Can you explain what you mean? 
You want _one_ worker to process the subscription from start?
That isn't possible. All the workers share the subscription.
You can create _another_ subscription for that.

However, I don't understand the context for this need, can you explain?

**Reply by tehkyle** [reply]
Ah, I think I've misunderstood the concurrent worker feature.
It's design is to allow multiple clients to consume the same subscription, but not wholely independently? 
The docs on concurrent worker connection failure seem to indicate this: "When a concurrent worker's connection ends unexpectedly, the server may assign the documents this worker has been processing to any other concurrent worker that is available."

Though in the examples (under "Run both workers" in docs) each concurrent worker has separate batch processing code blocks. Does this mean that each worker could process batches entirely differently, and be getting separate subsets of the entire subscription query stream?

We have a usage of data subscriptions that hooks up a set of queries to a particular target. That target can then be configured to process the data streams in specific ways. We were interested in concurrent workers because we don't have a defined limit to the number of targets that could be configured in the product and thought the feature would allow us to attach each target as a worker to a common set of query subscriptions so as not to bloat the total number of subscriptions, but (and correct me if I'm wrong) this seems not to be the case.

The subscription documentation states that subscriptions use Change Vectors to keep persistent progress, but the above documentation about concurrent workers would indicate that the change vector is for the subscription itself and not for each worker, correct?

Thanks,
Kyle

**Reply by ayende** [reply]
A concurrent subscription allows a set of workers to cooperate on processing the documents in a subscription.

Let's assume you have documents: `[1,2,3,4,5,6,7,8,9,10]` and two workers (batch size = 2).

The way it will work is so:

* Worker A - `[1,2]`
* Worker B - `[3,4]`
* Worker B - Batch done
* Worker B - `[5,6]`
* Worker A - Batch done
* Worker A - `[7,8]`
* Worker B - failed
* Worker A - `[5,6]` - taking over previously processed documents because of failures


The idea is that they share the workload, not reading from the same stream.

Think: queue, not a log stream like kafka.

For what you want, better to use separate subscriptions.

**Reply by tehkyle** [reply]
Got it, thank you.
Just to clarify the example code for concurrent workers as well, they are not enforced to perform the same work across Worker A and Worker B, is that correct? They each have a separate processing code block, so is it up to the implementation to ensure that the shared work isn't arbitrarily different?

**Reply by ayende** [reply]
Yes, that is correct. It doesn't really make sense to have them execute different things, but you can.
